_posts/2019-01-03-elixir-kaffe-codealong.md

```diff
@@ -0,0 +1,204 @@
+
+This post assumes basic knowledge of Elixir and no knowledge of Kafka or Kaffe. Here is the repo with the full project: [Elixir Kaffe Codealong](https://github.com/meryldakin/elixir_kaffe_codealong).
+
+## What is Kafka, briefly?
+Kafka is a messaging system. It does essentially three things:
+1. Receives messages from applications
+2. Keeps those messages in the order they were receieved in
+3. Allows other applications to read those messages in order
+
+Learn a lot more here: [Kafka in a Nutshell](https://sookocheff.com/post/kafka/kafka-in-a-nutshell/)
+
+## Basic Kafka terminology in simple terms
+- **consumer:** what is receiving messages from Kafka
+- **producer:** what is sending messages to Kafka
+- **topic:** a way to organize messages and allow consumers to only subscribe to the ones they want to receive
+- **partition:** allows a topic to be split among multiple machines and retain the same data so that more than one consumer can read from a single topic at a time
+- **leader/replica:** these are types of partitions. There is one leader and multiple replicas. The leader makes sure the replicas have the same and newest data. If the leader fails, a replica will take over as leader.
+- **offset:** the unique identifier of a message that keeps its order within Kafka
+
+## Codealong: basic Elixir app & Kafka running locally
+
+### Set up Kafka Server
+Follow the first two steps of the [quickstart instructions](http://kafka.apache.org/documentation/#quickstart) from Apache Kafka:
+1. [Download the code](https://www.apache.org/dyn/closer.cgi?path=/kafka/2.1.0/kafka_2.11-2.1.0.tgz)
+2. Start the servers
+Zookeeper
+`bin/zookeeper-server-start.sh config/zookeeper.properties`
+Kafka
+`bin/kafka-server-start.sh config/server.properties`
+
+### Set up Elixir App
+
+**1. Start new project**
+`mix new elixir_kaffe_codealong`
+
+**2. Configure kaffe**
+**2.a:** In `mix.exs` add `:kaffe` to the list of extra applications:
+```elixir
+def application do
+  [
+    extra_applications: [:logger, :kaffe]
+  ]
+end
+```
+
+**2.b:** Add kaffe to list of dependencies:
+```elixir
+defp deps do
+  [
+    {:kaffe, "~> 1.9"}
+  ]
+end  
+```
+
+**2.c:** Run `mix deps.get` in the terminal to lock new dependencies.
+
+**3. Configure producer**
+in `config/config.exs` add:
+```elixir
+config :kaffe,
+  producer: [
+    endpoints: [localhost: 9092],
+    # endpoints references [hostname: port]. Kafka is configured to run on port 9092.
+    # In this example, the hostname is localhost because we've started the Kafka server
+    # straight from our machine. However, if the server is dockerized, the hostname will
+    # be called whatever is specified by that container (usually "kafka")
+    topics: ["our_topic", "another_topic"], # add a list of topics you plan to produce messages to
+  ]
+```
+
+**4. Configure consumer**
```

Comment:
I believe we need a newline after this

![image](https://user-images.githubusercontent.com/73386/50737751-fa5b7400-1189-11e9-83c4-3fd98c79a7aa.png)